social-network-api/src/main/java/com/socialnetwork/api/controller/PostController.java

```diff
@@ -1,90 +1,137 @@
+    Integer originalPostId = postDto.getOriginalPostId();
+    Post post = convertToPost(postDto, user, originalPostId);
+    String header = request.getHeader(AUTHORIZATION_HEADER);
+
+    jwtTokenUtil.verifyUsernames(header, user.getUsername());
     postService.save(post);
-    return ResponseEntity.status(201).body(new Response("Post was created"));
+    return new Response("Post was created");
   }
 
-  //  @PatchMapping()
-  //  public ResponseEntity<?> editPost(@RequestBody PostDto postDto) {
-  //    Post post = convertToPost(postDto);
-  //    if (!postService.existsById(post.getPostId())) {
-  //      return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new Response(POST_NOT_FOUND));
-  //    }
-  //    postService.edit(post);
-  //    return ResponseEntity.ok(new Response("Post was edited successfully"));
-  //  }
-
-  //  @DeleteMapping()
-  //  public ResponseEntity<?> deletePost(@RequestBody PostDto postDto) {
-  //    Post post = convertToPost(postDto);
-  //    if (!postService.existsById(post.getPostId())) {
-  //      return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new Response(POST_NOT_FOUND));
-  //    }
-  //    postService.delete(post);
-  //    return ResponseEntity.ok(new Response("Post was deleted"));
-  //  }
-
-  private Post convertToPost(PostDto postDto) {
-    return modelMapper.map(postDto, Post.class);
+  @PostMapping("/bookmark")
+  public PostDto.Response.Default
+    saveBookmark(@RequestBody PostDto.Request.Action postDto, HttpServletRequest request)
+          throws NoUserWithSuchCredentialsException, NoPostWithSuchIdException, AccessDeniedException {
+    User user = userService.getReferenceById(postDto.getUser().getId());
+    Post post = postService.getReferenceById(postDto.getPost().getId());
+    String header = request.getHeader(AUTHORIZATION_HEADER);
+
+    jwtTokenUtil.verifyUsernames(header, user.getUsername());
+
+    if (bookmarkService.existsByUserAndPost(user, post)) {
+      bookmarkService.delete(user, post);
+    } else {
+      bookmarkService.save(user, post);
+    }
+
+    return convertToPostDto(post);
+  }
+
+  @PostMapping("/like")
+  public PostDto.Response.Default
+    saveLike(@RequestBody PostDto.Request.Action postDto, HttpServletRequest request)
+          throws NoUserWithSuchCredentialsException, NoPostWithSuchIdException, AccessDeniedException {
+    User user = userService.getReferenceById(postDto.getUser().getId());
+    Post post = postService.getReferenceById(postDto.getPost().getId());
+    String header = request.getHeader(AUTHORIZATION_HEADER);
+
+    jwtTokenUtil.verifyUsernames(header, user.getUsername());
+
+    if (likeService.existsByUserAndPost(user, post)) {
+      likeService.delete(user, post);
+    } else {
+      likeService.save(user, post);
+    }
+
+    return convertToPostDto(post);
+  }
+
+  private Post convertToPost(PostDto.Request.Created postDto, User user, Integer originalPostId)
```

Comment:
don`t pass originalpostid here. Just check if postdto`s original post exists, if not, throw an exception, if yes, map it, There is no need to save it if the request data is incorrect.